File: lua/pe004.lua
-- this computes projecteuler.net problem 004

local MAX3 = 999
local MIN3 = 100
local maxp = 0

local mini = MIN3
local minj = MIN3
local maxj = MAX3

local product = 0

local function is_palindromic(digits)
  slen = string.len(digits)

  for i = 1, slen - 1 do
    if string.byte(digits, i) ~= string.byte(digits, -i) then
      return false
    end
  end

  return true
end
  
for i = MAX3, mini, -1 do
  for j = maxj, minj, -1 do
    product = i * j
    if is_palindromic(tostring(product)) then
      if product > maxp then
        maxp = product
        -- minj = product 
        mini = product -- 1000
      end
    end
  end
  maxj = i
end

print(maxp)

File: python/pe004.py
#!/usr/bin/env python3

""" this computes projecteuler.net problem 004 """

import sys


def is_palindromic(digits):
    """This function returns True for palindromic input"""
    slen = len(digits)

    for k in range(slen):
        if digits[k] != digits[slen - (k + 1)]:
            return False

    return True


MAX3 = 999
MIN3 = 100

maxp = 0
mini = MIN3
minj = MIN3
maxj = MAX3
for i in range(MAX3, mini, -1):
    for j in range(maxj, minj, -1):
        product = i * j
        if is_palindromic(str(product)):
            if product > maxp:
                maxp = product
                minj = product // 1000
                mini = product // 1000
    maxj = i

print(maxp)

sys.exit(0)
